File: gamey/board.py
from enum import Enum

class MoveResult(Enum):
    SUCCESS = 0
    OCCUPIED = 1
    VICTORY = 2

target = {'A', 'B', 'C'}

class Node:
    def __init__(self, sides: set = set()):
        self.color: str = None
        self.sides: set = sides
        self.neighbors: list[Node] = []

    def add_neighbor(self, node: 'Node'):
        self.neighbors.append(node)
        node.neighbors.append(self)

    def __get_same_color_neighbors(self) -> list['Node']:
        return list(filter(lambda v: v.color == self.color, self.neighbors))

    def __spread_to_neighbors(self, neighbors: list['Node']):
        for v in neighbors:
            if v.sides < self.sides:
                v.sides = v.sides.union(self.sides)
                other_neighbors = v.__get_same_color_neighbors()
                other_neighbors.remove(self)
                v.__spread_to_neighbors(other_neighbors)

    def __recalculate_connected_sides(self):
        neighbors = self.__get_same_color_neighbors()
        for v in neighbors:
            self.sides = self.sides.union(v.sides)
        self.__spread_to_neighbors(neighbors)

    def set_color(self, color: str):
        self.color = color
        self.__recalculate_connected_sides()

class Board:
    def __init__(self, size: int):
        assert (size > 1)

        self.rows: list[list[Node]] = []
        for i in range(size):
            self.rows.append([])
            for j in range(i+1):
                s = set()
                if(j == 0):      s.add('A')
                if(i == j):      s.add('B')
                if(i == size-1): s.add('C')

                n = Node(s)
                if(i > j):
                    n.add_neighbor(self.rows[i-1][j])
                if(j > 0):
                    n.add_neighbor(self.rows[i][j-1])
                    n.add_neighbor(self.rows[i-1][j-1])
                self.rows[i].append(n)

    def place_piece(self, i: int, j: int, color: str):
        node: Node = self.rows[i][j]
        if(node.color != None):
            return MoveResult.OCCUPIED
        node.set_color(color)
        if node.sides == target:
            return MoveResult.VICTORY
        return MoveResult.SUCCESS

File: gamey/board.ts
export enum MoveResult {
    SUCCESS = 0,
    OCCUPIED = 1,
    VICTORY = 2,
}

const target = new Set<string>(['A', 'B', 'C']);

export class Node {
    color: string = '';
    neighbors: Node[] = []
    public constructor(
        public sides: Set<String>,
    ) { }

    public addNeighbor(node: Node) {
        this.neighbors.push(node)
        node.neighbors.push(this)
    }

    private getSameColorNeighbors(): Node[] {
        return this.neighbors.filter((v) => v.color == this.color)
    }

    private spreadToNeighbors(neighbors: Node[]) {
        for (const v of neighbors) {
            if ([...v.sides].every((side) => this.sides.has(side))) {
                v.sides = new Set<String>([...v.sides, ...this.sides])
                const otherNeighbors = v.getSameColorNeighbors()
                    .filter(vv => vv !== this)
                v.spreadToNeighbors(otherNeighbors)
            }
        }
    }

    private recalculateConnectedSides() {
        const neighbors = this.getSameColorNeighbors()
        for (const v of neighbors) {
            this.sides = new Set<String>([...this.sides, ...v.sides])
        }
        this.spreadToNeighbors(neighbors)
    }

    public setColor(color: string) {
        this.color = color
        this.recalculateConnectedSides()
    }
}

export class Board {
    rows: Node[][] = []
    public constructor(size: number) {
        if (size <= 1) throw new Error();

        for (let i = 0; i < size; ++i) {
            this.rows.push([])
            for (let j = 0; j <= i; ++j) {
                // const s = new Set<string>([
                //     ...(j == 0 ? ['A'] : [])
                // ])
                const s = new Set<string>()
                if (j == 0) s.add('A');
                if (i == j) s.add('B');
                if (i == size - 1) s.add('C');

                const n = new Node(s)
                if (i > j) {
                    n.addNeighbor(this.rows[i - 1][j]);
                }
                if (j > 0) {
                    n.addNeighbor(this.rows[i][j - 1]);
                    n.addNeighbor(this.rows[i - 1][j - 1]);
                }
                this.rows[i].push(n)
            }
        }
    }

    public placePiece(i: number, j: number, color: string): MoveResult {
        const node = this.rows[i][j]
        if (node.color !== '') {
            return MoveResult.OCCUPIED
        }
        node.setColor(color)
        if ([...target].every(side => node.sides.has(side))) {
            return MoveResult.VICTORY
        }
        return MoveResult.SUCCESS
    }
}

